GET_ORGANIZATION permission for Env teams

SofiaSazonova opened the pull request:
### Feature or Bugfix
- Bugfix


### Detail
-  When Environment is linked, check that SamlGroup is invited to Organization
- When group is invited to Env, add GET_ORGANIZATION permissions, if it doesn't exist yet
- When group is deleted from Env Teams, check 1)it's not invited to Org, 2)it's not invited to another Env of this Org -> remove GET_ORGANIZATION permissions
- Related tests
- list_organisations now is based on GET_ORGANIZATION permissions, so non-mebers with GET_ORGANIZATION permissios can see Orgs 
- Inside OrgView all buttons (link, Invite, update, delete) are hidden basing on permissions
- migration script for permission backfill

### Relates
- #1147

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

SofiaSazonova commented:
Checked in AWS:
Org, Env1, Env2, Group (not in Org)

- [ ] Can't create an Env with Group
- [ ] Can create Env with group that is invited into Org

- [ ] When Group is invited to the Env1, it gets GET_ORGANIZATION permissions for Org
- [ ] When Group is invited to the Env2, GET_ORGANIZATION permission for Org is not duplicated

- [ ] When Group is deleted from Env2,  GET_ORGANIZATION permission for Org is NOT deleted
- [ ] When Group is deleted from Env1,  GET_ORGANIZATION permission for Org is deleted

- [ ] Group is invited into Org, then invited into Env1, GET_ORGANIZATION permission for Org is not duplicated
- [ ] When Group is deleted from Env1,  GET_ORGANIZATION permission for Org is NOT deleted
- [ ] When Group is deleted from Org,  GET_ORGANIZATION permission for Org is deleted

noah-paige commented on backend/dataall/core/environment/services/environment_service.py:
```diff
@@ -213,6 +215,12 @@ def create_environment(uri, data=None):
         context = get_context()
         with context.db_engine.scoped_session() as session:
             EnvironmentRequestValidationService.validate_creation_params(data, uri, session)
+            if not OrganizationRepository.is_group_invited(session, uri, data['SamlGroupName']):
+                raise Exception(
+                    f'Group {data["SamlGroupName"]} is not a member of the organization {data.get("organizationUri")}. '
+                    f'Invite this group to the organisation before linking the environment.'
+                )
+
```
Can we move to to part of `validate_user_groups()` in `EnvironmentRequestValidationService`

SofiaSazonova replied:
done

noah-paige commented on backend/dataall/core/organizations/db/organization_repositories.py:
```diff
@@ -106,6 +106,19 @@ def query_organization_groups(session, uri, filter) -> Query:
+                )
+            )
+            .count()
+        ) > 0
```
nit: Don't need to specify `>0` - if returns a value greather than 0 will evaluate as True in the if condition either way

noah-paige commented on backend/dataall/core/environment/db/environment_repositories.py:
```diff
@@ -303,6 +303,22 @@ def is_user_invited_to_environment(session, groups, uri):
+                )
+            )
+            .count()
+            > 0
```
nit: Don't need to specify >0 - if returns a value greather than 0 will evaluate as True in the if condition either way

noah-paige commented on backend/dataall/core/environment/services/environment_service.py:
```diff
@@ -463,6 +487,20 @@ def remove_group(uri, group):
                 resource_type=Environment.__name__,
             )
+
+            # if group is no longer invited to organization or environments inside it, remove GET_ORGAISATION policy
```
nit: type in comment `remove GET_ORGANIZATION policy`

SofiaSazonova replied:
done

noah-paige commented on backend/dataall/core/environment/services/environment_service.py:
```diff
@@ -409,6 +417,22 @@ def invite_group(uri, data=None) -> (Environment, EnvironmentGroup):
+                resource_type=Organization.__name__,
+                permission_name=GET_ORGANIZATION,
+            ):
+                ResourcePolicyService.attach_resource_policy(
+                    session=session,
+                    group=group,
+                    resource_uri=environment.organizationUri,
+                    permissions=[GET_ORGANIZATION],
+                    resource_type=Organization.__name__,
+                )
```
The way `def attach_resource_policy()` is written is that the policy is only created and permissions associated if and only if they do not already exist (like a "create if not exists" expressions)

Open ended Q - is there added benefit to having the check `if check_group_resource_permission()`?

dlpzx replied:
True, attach_resource_policy does not create duplicates if the policy already exists. I don't think we need the check at creation, only at deletion

SofiaSazonova replied:
done

noah-paige commented:
Also a note - I believe there is a bug with the confirmation pop ups when removing an invited team or consumption role from an environment

Let's say I have 3 groups in an env (GroupA, B, and C)... When removing GroupA - I receive the pop up and need to type `permanently delete` for the group
- Once I do so GroupA is removed but the pop up remains and the team it is asking me to confirm to remove now is GroupB 

In `EnvironmentTeams.js` we have one variable `isDeleteTeamModalOpen` to control if the mondals for N groups is open - some bug in the logic is being introduced because of this

SofiaSazonova commented:
@noah-paige 
(1) I think, it's for separate PR
(2) Done. I didn't remove REMOVE_ORGANIZATION_GROUP yet, but backfilled GET_ORGANISATION
(3) I changed OrganisationView. Now readers are displayed. Also, I changed list_organisation logic to show Orgs for "readers"
(4) I will fix this bug in separate PR

SofiaSazonova commented:
![image](https://github.com/data-dot-all/dataall/assets/1146495/497aee7e-b541-4f0a-bf7d-e3105f632be2)

noah-paige commented on backend/dataall/core/groups/services/group_service.py:
```diff
@@ -72,6 +72,6 @@ def list_groups_without_invited(filter: dict = None):
             for group in cognito_groups:
                 if group not in invited_groups:
-                    groups.append({'groupName': group})
+                    groups.append({'groupUri': group})
```
why this change?

noah-paige commented on backend/dataall/core/groups/api/types.py:
```diff
@@ -45,6 +45,18 @@
+    ],
+)
+
+OrganizationReadersSearchResult = gql.ObjectType(
+    name='OrganizationReaders',
+    fields=[
+        gql.Field(name='count', type=gql.Integer),
+        gql.Field(name='page', type=gql.Integer),
+        gql.Field(name='pages', type=gql.Integer),
+        gql.Field(name='hasNext', type=gql.Boolean),
+        gql.Field(name='hasPrevious', type=gql.Boolean),
+        gql.Field(name='nodes', type=gql.ArrayType(GroupsInfo)),
```
Is this type used anywhere?

noah-paige commented on backend/dataall/core/groups/api/types.py:
```diff
@@ -45,6 +45,18 @@
     name='GroupsInfo',
     fields=[
-        gql.Field(name='groupName', type=gql.String),
+        gql.Field(name='groupUri', type=gql.String),
```
why change this from groupName to groupUri?

noah-paige commented on backend/dataall/core/organizations/db/organization_repositories.py:
```diff
@@ -106,6 +122,52 @@ def query_organization_groups(session, uri, filter) -> Query:
+                Permission.permissionUri == ResourcePolicyPermission.permissionUri,
+            )
+            .filter(
+                or_(
```
don't need the `or_(...)` condition

noah-paige commented on frontend/src/modules/Organizations/components/OrganizationTeamInviteForm.js:
```diff
@@ -48,8 +48,8 @@ export const OrganizationTeamInviteForm = (props) => {
             ...g,
-            value: g.groupName,
-            label: g.groupName
+            value: g.groupUri,
```
similar to above - any need to change `groupName` to `groupUri`?

noah-paige commented on backend/migrations/versions/11b716ddb036_backfill_get_org_permissions.py:
```diff
@@ -0,0 +1,50 @@
+
+
+# revision identifiers, used by Alembic.
+revision = '11b716ddb036'
+down_revision = '458572580709'
```
I see the migration script is failing because of multiple heads in alembic ... I believe you will need to merge the latest from `main` to this branch and then update the `down_revision` here to be pointing from the latest revision currently in main

noah-paige commented:
The new frontend UI looks great! 1 general question - does the query still run in manageable time for either `getOrganization` or `listOrganizationReadOnlyGroups` with the nested resolvers / added logic?

I was considering the idea that we could add a new column to `OrganizationGroup` called `userType` and backfiling for existing Groups. Then for each invite or remove from envs we also create/remove a new OrganizationGroup record with `userType` of value `Reader` 
- Could make it easier to determine what should be loaded for a given user on UI / when listing reader vs admin users 

... but this is not really needed unless we see some big performance differences I would argue

SofiaSazonova commented:
Well, right now I can't spot any big difference. We have queries like this in many places. But of course, it's better to test with some loaded infrastructure.

As for OrganisationGroups. Well, we can do this, but it means, that we also need to redo the UI and backend of inviting groups: we want then to add opportunity to give Readonly permissions. 
May be, it's a good thing to do. I'm not sure right now. Let's discuss

noah-paige commented:
Last required change I think - the resolved user role for an invite EnvGroup (i.e. an Org Reader but not an Administrator) shows as `NotMember`

![Screenshot 2024-05-24 at 10 59 34 AM](https://github.com/data-dot-all/dataall/assets/69586985/de7e0d68-5f91-442b-a810-e82e301e560a)

Would need to update the way `find_organization_membership` identifies who is part of org group or not

noah-paige commented on backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:
```diff
@@ -108,9 +108,9 @@ def find_resource_policy(session, group_uri: str, resource_uri: str, resource_ty
     @staticmethod
     def find_all_resource_policies(
-        session, group_uri: str, resource_uri: str, resource_type: str = None, permissions: List[str] = None
+        session, groups: [str], resource_uri: str, resource_type: str = None, permissions: List[str] = None
```
nit: `List[str]`

noah-paige reviewed (approved):
Left 2 minor comments but overall tested and looks good
